### PR TITLE
Run post-upgrade actions for all requests

### LINKED ISSUE #2153: Post-upgrade actions are only ran in wp-admin
I am currently working on #1991 which will be running some code after WCPay is upgraded.

I noticed [there is already an existing mechanism to execute post-upgrade actions](https://github.com/Automattic/woocommerce-payments/blob/d71dbaa2842c9e0320db10696cacceda4b20d5bc/includes/class-wc-payments.php#L696) so I wanted to use it. The only other callback currently hooked into the `woocommerce_woocommerce_payments_updated` action is `WC_Payments_Apple_Pay_Registration::verify_domain_if_configured()`.

However, I noticed that the `woocommerce_woocommerce_payments_updated` action is [ran during `admin_init`](https://github.com/Automattic/woocommerce-payments/blob/d71dbaa2842c9e0320db10696cacceda4b20d5bc/includes/class-wc-payments.php#L144), meaning that it only runs when an admin page is visited (aside from some other possible cases - `admin-ajax.php`?). For #1991, it is important that the post-upgrade actions are always performed after the plugin is upgraded, not only when wp-admin is visited.

To check if this is actually an issue, I did some testing and noticed that the `admin_init` hook is not fired e.g. when upgrading a plugin using WP-CLI, meaning this approach isn't reliable enough as an "always executed after upgrade" solution.

I also did some digging to find out if there was a specific reason why `admin_init` was chosen. Tracking down [the PR where this was implemented in WCPay](https://github.com/Automattic/woocommerce-payments/pull/1239) I found references to copying the logic from `woocommerce-gateway-stripe`. In the latter's repo, [I didn't find any references for a possible reason for this.](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1054)

My _guess_ is that the reason could have been that plugin upgrades are ran through wp-admin and this hook was used so it's not ran unnecessarily during non-admin browsing? 🤷

Before digging into it more or changing anything, I thought I'd ask if anyone is familiar with the reason why it was done this way.

For #1991, I'd like to execute the code during `plugins_loaded` or `init`. I am wondering if this could cause issues if the Apple Pay domain verification happened during these hooks instead of on `admin_init`.

One drawback of this that comes to mind would be that the very first front-end request after an upgrade could be slower due to the extra operations which include a `wp_remote_get()` call (albeit to the site itself). The benefit of always running post-upgrade actions (as opposed to only after accessing wp-admin) and having just one post-upgrade hook for all kinds of upgrades instead of just wp-admin ones seems worth it for me.

### PR DESCRIPTION
**Pull request by waclawjacek**
Fixes #2153.

#### Changes proposed in this Pull Request

While working on #1991, I noticed that the post-upgrade hook `woocommerce_woocommerce_payments_updated` only runs when visiting a wp-admin page.

Because the migrations in #1991 have to run before any front-end or wp-admin request, this PR introduces a change to the upgrade behavior, running them on `init` rather than on `admin_init`.

It is also better to run this migration during a non-admin request because if the plugin is upgraded e.g. using WP-CLI, the migrations wouldn't run.

Kudos to @ricardo and @DanReyLop for the insight and great discussion on how to tackle this in p1623435917484000-slack-CGGCLBN58 🙌

#### Testing instructions

As currently the only module (other than the new code implemented in #1991) using `woocommerce_woocommerce_payments_updated` is the Apple Pay domain association, let's use it for testing to make sure there are no regressions.

* Set up your site so you can use Apple Pay ([testing using live credentials is a requirement](https://docs.woocommerce.com/document/payments/apple-pay/#troubleshooting)).
* Make sure you enable payment request buttons.

**To verify previous behavior:**

1. Check out an earlier commit prior to this PR.
2. A part of the Apple Pay verification procedure is a file being created on the server. Remove the file `/.well-known/apple-developer-merchantid-domain-association` from your test site.
3. Manually change the plugin version in the database to an earlier version, e.g. `2.4.0`:
    `UPDATE wp_options SET option_value = '2.4.0' WHERE option_name = 'woocommerce_woocommerce_payments_version'`
4. Visit a front-end page. The `/.well-known/apple-developer-merchantid-domain-association` file should still be missing.
5. Visit any wp-admin page. The above file should now be present.

**To verify new behavior:**

Do the same as above, but in the "Visit a front-end page" step, the file should be there.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

File: includes/class-wc-payments-apple-pay-registration.php [hunk 1]
```diff
@@ -79,16 +83,12 @@ public function __construct( WC_Payments_API_Client $payments_api_client, WC_Pay
 	 * @return  void
 	 */
 	public function init() {
-		$this->gateway = WC_Payments::get_gateway();
```
Comment (frosso) [inline includes/class-wc-payments-apple-pay-registration.php:82]: Nice improvement!

File: includes/class-wc-payments-apple-pay-registration.php [hunk 2]
```diff
@@ -61,15 +61,19 @@ class WC_Payments_Apple_Pay_Registration {
...
 	 */
-	public function __construct( WC_Payments_API_Client $payments_api_client, WC_Payments_Account $account ) {
+	public function __construct( WC_Payments_API_Client $payments_api_client, WC_Payments_Account $account, WC_Payment_Gateway_WCPay $gateway ) {
 		$this->domain_name             = $_SERVER['HTTP_HOST'] ?? str_replace( [ 'https://', 'http://' ], '', get_site_url() ); // @codingStandardsIgnoreLine
 		$this->apple_pay_verify_notice = '';
 		$this->payments_api_client     = $payments_api_client;
 		$this->account                 = $account;
+		$this->gateway                 = $gateway;
 
+		add_action( 'init', [ $this, 'add_domain_association_rewrite_rule' ], 5 );
```
Comment (frosso) [inline includes/class-wc-payments-apple-pay-registration.php:75]: I see that this has been moved from `init` to its own action with a higher priority. Did the previous code not work well?

**Review by frosso — APPROVED** [review]
Changes look great :shipit: just one question out of curiosity

File: includes/class-wc-payments-apple-pay-registration.php (hunk 2 above)
Comment (waclawjacek) [reply includes/class-wc-payments-apple-pay-registration.php:75]: It did, but the upgrade hook also runs during `init` at the same priority - 10, so we need to make sure the rewrite rules are added prior to that (just like they were when the entire code was in one method).

Comment (frosso) [reply includes/class-wc-payments-apple-pay-registration.php:75]: I see, that makes perfect sense! Thanks for the explanation! 💯

**Pull request merged**